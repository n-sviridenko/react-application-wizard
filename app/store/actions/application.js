import { createAction } from 'redux-actions';

export const SHOW_REQUEST = 'app/application/SHOW_REQUEST';
export const SHOW_SUCCESS = 'app/application/SHOW_SUCCESS';
export const SHOW_FAIL = 'app/application/SHOW_FAIL';

export const showRequest = createAction(SHOW_REQUEST, null, (id) => ({ key: id }));
export const showSuccess = createAction(SHOW_SUCCESS, null, (item, id) => ({ key: id }));
export const showFail = createAction(SHOW_FAIL, null, (err, id) => ({ key: id }));

export const CREATE_REQUEST = 'app/application/CREATE_REQUEST';
export const CREATE_SUCCESS = 'app/application/CREATE_SUCCESS';
export const CREATE_FAIL = 'app/application/CREATE_FAIL';

export const createRequest = createAction(CREATE_REQUEST);
export const createSuccess = createAction(CREATE_SUCCESS);
export const createFail = createAction(CREATE_FAIL);
