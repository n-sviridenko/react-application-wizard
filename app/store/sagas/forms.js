import { SubmissionError } from 'redux-form/immutable';
import { take, takeEvery, race, put, call } from 'redux-saga/effects';

import { SUBMIT } from 'store/actions/forms';

export function* formSubmit({ meta }) {
  const {
    requestAction,
    successActionType,
    failActionType,
    resolve,
    reject,
  } = meta;

  yield put(requestAction);

  const { success } = yield race({
    success: take(successActionType),
    failure: take(failActionType),
  });

  if (success) {
    yield call(resolve, success.payload);
  } else {
    yield call(reject, new SubmissionError());
  }
}

export function* formSubmitWatcher() {
  yield takeEvery(SUBMIT, formSubmit);
}

export default [
  formSubmitWatcher,
];
