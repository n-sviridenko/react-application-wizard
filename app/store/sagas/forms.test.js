/* eslint-disable redux-saga/yield-effects */
import { SubmissionError } from 'redux-form';
import { take, takeEvery, race, call } from 'redux-saga/effects';

import { SUBMIT } from 'store/actions/forms';
import { formSubmit, formSubmitWatcher } from './forms';

describe('forms saga', () => {
  const requestAction = {
    type: 'TEST_REQUEST',
    payload: { foo: 'bar' },
  };

  const successActionType = 'TEST_SUCCESS';
  const failActionType = 'TEST_FAIL';
  const resolve = () => {};
  const reject = () => {};

  let formSubmitTask;

  beforeEach(() => {
    formSubmitTask = formSubmit({
      meta: {
        requestAction,
        successActionType,
        failActionType,
        resolve,
        reject,
      },
    });

    const requestDescriptor = formSubmitTask.next().value;
    expect(requestDescriptor).toMatchSnapshot();

    const raceDescriptor = formSubmitTask.next().value;
    const expectedRaceDescriptor = race({
      success: take(successActionType),
      failure: take(failActionType),
    });
    expect(raceDescriptor).toEqual(expectedRaceDescriptor);
  });

  it('should execute resolve it successAction was received', () => {
    const actionDescriptor = formSubmitTask.next({ success: { payload: {} } }).value;
    expect(actionDescriptor).toEqual(call(resolve, {}));
  });

  it('should execute resolve it failAction was received', () => {
    const actionDescriptor = formSubmitTask.next({ failure: { payload: {} } }).value;
    expect(actionDescriptor).toEqual(call(reject, new SubmissionError({})));
  });
});

describe('formSubmitWatcher Saga', () => {
  const formSubmitWatcherSaga = formSubmitWatcher();

  it('should start task to watch for SUBMIT action', () => {
    const takeLatestDescriptor = formSubmitWatcherSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeEvery(SUBMIT, formSubmit));
  });
});
