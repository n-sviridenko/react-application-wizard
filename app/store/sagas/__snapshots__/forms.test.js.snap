exports[`forms saga should execute resolve it failAction was received 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "foo": "bar",
      },
      "type": "TEST_REQUEST",
    },
    "channel": null,
  },
}
`;

exports[`forms saga should execute resolve it successAction was received 1`] = `
Object {
  "@@redux-saga/IO": true,
  "PUT": Object {
    "action": Object {
      "payload": Object {
        "foo": "bar",
      },
      "type": "TEST_REQUEST",
    },
    "channel": null,
  },
}
`;
